service/session.hh

```diff
@@ -0,0 +1,121 @@
+
+constexpr session_id default_session_id = session_id::create_null_id();
+
+/// Session is used to track execution of work related to some greater task, identified by session_id.
+/// Work can enter the session using enter(), and is considered to be part of the session
+/// as long as the guard returned by enter() is alive.
+///
+/// Session goes over the following states monotonically:
+///  1) open - accepts work via enter()
+///  2) closing - rejects work via enter()
+///  3) closed - rejects work via enter(), and no guards are alive anymore
+///
+/// Sessions are removed only after they are closed, it's impossible to have a session::guard of a session
+/// which is not in the registry.
+class session {
+public:
+    using link_type = bi::list_member_hook<bi::link_mode<bi::auto_unlink>>;
+private:
+    session_id _id;
+    seastar::gate _gate;
+    std::optional<shared_future<>> _closed;
+    link_type _link;
+public:
+    using list_type = boost::intrusive::list<session,
+            boost::intrusive::member_hook<session, session::link_type, &session::_link>,
+            boost::intrusive::constant_time_size<false>>;
+public:
+    class guard {
+        session* _session;
+        seastar::gate::holder _holder;
+    public:
+        guard(session& s) : _session(&s), _holder(s._gate.hold()) {}
+
+        void check() {
+            if (_session->_closed) {
+                throw seastar::abort_requested_exception();
+            }
+        }
+    };
+
+    session(session_id id) : _id(id) {}
+
+    guard enter() {
+        return guard(*this);
+    };
+
+    /// No new work is admitted to enter the session after this.
+    /// Can be called many times.
+    void start_closing() noexcept {
+        if (!_closed) {
+            _closed = seastar::shared_future<>(_gate.close());
+        }
+    }
+
+    /// Returns true iff the session is not open.
+    /// Calling enter() in this state will fail.
+    bool is_closing() const {
+        return bool(_closed);
+    }
+
+    session_id id() const {
+        return _id;
+    }
+
+    /// Post-condition of successfully resolved future: There are no guards alive for this session, and
+    /// and it's impossible to create more such guards later.
+    /// Can be called concurrently.
+    future<> close() {
+        start_closing();
+        return _closed->get_future();
+    }
+};
+
+class session_manager {
+    std::unordered_map<session_id, std::unique_ptr<session>> sessions;
+    session::list_type closing_sessions;
+    seastar::semaphore session_drain_sem{1};
+public:
+    session_manager();
+
+    session::guard enter_session(session_id id);
+
+    /// Creates a session on this shard if it doesn't exist yet.
+    /// If the session already exists does nothing.
+    void create_session(session_id);
+
+    /// Calls start_closing() on all sessions except those in keep.
+    void prune_sessions_except(const std::unordered_set<session_id>& keep);
```

Comment:
Why not just call it close_sessions_except()? "prune" makes it sound like we have more operations than closing.

Reply:
I'm fine with close(), but then @avikivity will say that close() implies that we wait, but here we only start closing. So `initiate_close_of_sessions_except()`?

Reply:
The new name sounds good.